test/hooks.on-listen.test.js

```diff
@@ -0,0 +1,974 @@
+'use strict'
+
+const { t, before } = require('tap')
+const Fastify = require('../fastify')
+const fp = require('fastify-plugin')
+const dns = require('dns').promises
+const split = require('split2')
+
+let localhost
+
+before(async function () {
+  const lookup = await dns.lookup('localhost')
+  localhost = lookup.address
+})
+
+t.test('localhost onListen should be called in order', t => {
+  t.plan(2)
+  const fastify = Fastify()
+  t.teardown(fastify.close.bind(fastify))
+  let order = 0
+
+  fastify.addHook('onListen', function (done) {
+    t.equal(order++, 0, '1st called in root')
+    done()
+  })
+
+  fastify.addHook('onListen', function (done) {
+    t.equal(order++, 1, '2nd called in root')
+    done()
+  })
+
+  fastify.listen({
+    host: 'localhost',
+    port: 0
+  })
+})
+
+t.test('localhost async onListen should be called in order', async t => {
+  t.plan(2)
+  const fastify = Fastify()
+  t.teardown(fastify.close.bind(fastify))
+  let order = 0
+
+  fastify.addHook('onListen', async function () {
+    t.equal(order++, 0, '1st async called in root')
+  })
+
+  fastify.addHook('onListen', async function () {
+    t.equal(order++, 1, '2nd async called in root')
+  })
+
+  await fastify.listen({
+    host: 'localhost',
+    port: 0
+  })
```

Comment:
```suggestion
  })
  t.equal(order, 0, 'the onListen hooks are not awaited')
```

Reply:
I tried using this. In our implementation, onListen hooks are not awaited because resolve is called independently of them. The problem is that the test you provide is not adequate because hooks can still be completed before reaching that point (after the await) since they are running in parallel. Thus, the order will not be asserted properly. Is there any other way we could prove they are not awaited?

Reply:
I added two unit tests at the end of the file. please look